Compression/src/CompressDetStepMCs_module.cc

```diff
@@ -0,0 +1,671 @@
+    for (auto& i_simParticle : *_newSimParticles) {
+      mu2e::SimParticle& newsim = i_simParticle.second;
+      const auto& newParentPtr = newsim.parent();
+      const auto& expectedParentPtr = _simPtrRemap.find(newParentPtr);
+      if (expectedParentPtr != _simPtrRemap.end()) {
+        newsim.parent() = expectedParentPtr->second;
+        if (_conf.debugLevel() > 0) {
+          std::cout << "Setting SimParticle (" << newsim.id() << ")'s parent to " << expectedParentPtr->second << std::endl;
+        }
+      }
+    }
+  }
+}
+
+void mu2e::CompressDetStepMCs::compressGenParticles() {
+  // Loop through the new SimParticles to keep any GenParticles
+  for (auto& i_simParticle : *_newSimParticles) {
+    mu2e::SimParticle& newsim = i_simParticle.second;
+    if(newsim.genParticle().isNonnull()) { // will crash if not resolvable
+      // Copy GenParticle to the new collection
+      _newGenParticles->emplace_back(*newsim.genParticle());
+      newsim.genParticle() = art::Ptr<mu2e::GenParticle>(_newGenParticlesPID, _newGenParticles->size()-1, _newGenParticleGetter);
+    }
+  }
+}
+
+void mu2e::CompressDetStepMCs::compressStepPointMCs(const art::Event& event) {
+
+  for (const auto& i_tag : _conf.stepPointMCTags()) {
+    const auto& stepPointMCs = event.getValidHandle<StepPointMCCollection>(i_tag);
+    if(_conf.debugLevel()>0 && stepPointMCs->size()>0) {
+      std::cout << "Compressing StepPointMCs from " << i_tag << std::endl;
+    }
+    for (const auto& stepPointMC : *stepPointMCs) {
+      if (_stepPointMCCompressionLevel == mu2e::CompressionLevel::kSimParticleCompression) {
+        for (const auto& simPartsToKeep : _simParticlesToKeep) {
+          const art::ProductID& oldProdID = simPartsToKeep.first;
+          if (stepPointMC.simParticle().id() != oldProdID) {
+            continue;
+          }
+          const SimParticleSet& alreadyKeptSimParts = simPartsToKeep.second;
+          for (const auto& alreadyKeptSimPart : alreadyKeptSimParts) {
+            if (stepPointMC.simParticle() == alreadyKeptSimPart) {
+              StepPointMC newStepPointMC(stepPointMC);
+              _newStepPointMCs.at(i_tag.instance())->push_back(newStepPointMC);
+            }
+          }
+        }
+      }
+      else if (_stepPointMCCompressionLevel == mu2e::CompressionLevel::kNoCompression) {
+        StepPointMC newStepPointMC(stepPointMC);
+        _newStepPointMCs.at(i_tag.instance())->push_back(newStepPointMC);
+        recordSimParticle(stepPointMC.simParticle());
+      }
+      else {
+        throw cet::exception("CompressDetStepMCs") << "Unrecognized compression level \"" << _stepPointMCCompressionLevel.name() <<"\" for StepPointMCs" << std::endl;
+      }
+    }
+  }
+}
+void mu2e::CompressDetStepMCs::compressMCTrajectories(const art::Event& event) {
+
+  const auto& mcTrajectories = event.getValidHandle<MCTrajectoryCollection>(_conf.mcTrajectoryTag());
```

Comment:
I forgot to mention another important thing.   The compression modules must allow some of their input products to be ignored, by e.g. explicitly specifying the empty string for the tag in fhicl config.  Otherwise these large modules will not be usable outside of a narrowly defined "production" job configuration.  I am attaching this comment to a line where the code will break if my file does not have trajectories saved, but this comment is more general and applies to other products and the other module as well.

Reply:
We are evolving to require common inputs for all jobs of a given type.  Fhicl is simply too limited to handle that internally, and the proliferation of scripts needed to deal inconsistent, incompatible inputs is unmaintainable.   What use case are you thinking of that requires this?

Reply:
I've added checks for empty tags as a way to do this in commit e59f3ca